ethexe/signer/src/digest.rs

```diff
@@ -0,0 +1,255 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2024 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Keccak256 digest type. Implements AsDigest hashing for ethexe common types.
+
+use core::fmt;
+use ethexe_common::router::{
+    BlockCommitment, CodeCommitment, OutgoingMessage, StateTransition, ValueClaim,
+};
+use gprimitives::{MessageId, H256};
+use parity_scale_codec::{Decode, Encode};
+use sha3::Digest as _;
+
+use crate::Address;
+
+#[derive(
+    Clone,
+    Copy,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Hash,
+    Encode,
+    Decode,
+    derive_more::From,
+    derive_more::Into,
+    derive_more::AsRef,
+)]
+pub struct Digest([u8; 32]);
+
+impl fmt::Debug for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+impl fmt::Display for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+/// Trait for hashing types into a Digest using Keccak256.
+pub trait AsDigest {
+    fn as_digest(&self) -> Digest;
```

Comment:
If we care about performance, then we should use [`Keccak256::update`](https://docs.rs/digest/0.10.6/digest/core_api/struct.CoreWrapper.html#impl-Update-for-CoreWrapper%3CT%3E) instead to avoid allocations

Reply:
Cannot understand how we can use it here. If you explain I would create an issue so that we can apply it soon

Reply:
Still can not see the point where we can use it and why this is correct

Reply:
Done